### ISSUE TITLE
Letters after numbers in body names used as abbreviations for weights & measures in certain voices

### ISSUE THREAD

**Issue by Darkcyde13** [cmdrmcdonald/EliteDangerousDataProvider#163]
Hi Cmdr McDonald,

I'm not sure if this is something you can 'fix', as it seems to depend on the voice used.

I've just been doing some exploring and came across this quirk in the speech output.  For planet designations ending in a G like 'PRO EURL IX-L D7-24 2 G' the speech is saying the G as 'grams'.

I've tried this with 'Ivona Amy' and 'MS Zira Desktop' voices and both say it as grams.  However, when I tried MS Hazel and MS David, they both say it correctly as G.

I'm not sure if you want to bother with something so small, particularly when it depends on the voice used, but I thought I'd report it anyway.

Thanks for your continued development of EDDI, and I can't wait to try the new version! :)

Kind regards,

Darkcyde.

EDIT:
I've also noticed that bodies (stars & planets) also sometimes say 'A' as 'uh' rather than 'ayy' when the A is at the end, like the G above.

...and I just scanned 'PRO EURL IX-L D7-24 2 H' and it said the H as 'hours'! :(

**Comment by Darkcyde13** [comment]
I've tested all the letters of the alphabet now.  With the two voices mentioned, when certain letters are spoken after a number, words are used instead. e.g. "11 t" is 'eleven tonnes'.  The following letters are used as these abbreviations:-

G = grams
H = hours
L = litres
M = meters
T = tonnes
Y = years

**Comment by Darkcyde13** [comment]
I've fixed this now.  I'm passing all instances of body and star names to a new function which corrects the spoken last letter so that the weights & measures are no longer said.

However, if you still choose to correct this internally, that would be great too. :)  I'll leave this open for a day or two in case you wish to mark it as a bug to be fixed.

**Comment by cmdrmcdonald** [comment]
How are you altering the letter?  It's something that I could add to EDDI, but there are a lot of similar items out there and it's quite voice-dependent (I can't tell a voice not to interpret the text so would have to hard-code each situation).

**Comment by Darkcyde13** [comment]
This is the function I quickly threw together.  You may be able to make it more efficient.  In the four events that call this (Body/Star Scanned and Body/Star Report), where the name was being said, I simply replaced it with a call to this function.


```
{_ BodyName}
{_ letters said as words after numbers}

{if reportbody.name:
   {set idx to len(reportbody.name)-2}
   {set chk to slice(reportbody.name, idx, 2)}
   {set part to slice(reportbody.name, 0, (idx))}
|elif event.name:
   {set idx to len(event.name)-2}
   {set chk to slice(event.name, idx, 2)}
   {set part to slice(event.name, 0, (idx))}
}

{if chk = " a" || chk = " A":
    {set chk to " ay"}
|elif chk = " g" || chk = " G":
    {set chk to " gee"}
|elif chk = " h" || chk = " H":
    {set chk to " aitch"}
|elif chk = " l" || chk = " L":
    {set chk to " elle"}
|elif chk = " m" || chk = " M":
    {set chk to " emm"}
|elif chk = " t" || chk = " T":
    {set chk to " tee"}
|elif chk = " y" || chk = " Y":
    {set chk to " why"}
|else:
    {set chk to ""}
}

{if chk = "":
   {P(reportbody.name)}
   {P(event.name)}
|else:
   {P(cat(part,chk))}
}

```

First I find the length of the body/star name and minus two from it.  This gives me the index I need to get the last two characters, and a total length of the part I need to keep.  I then check these to see if they match any of the letters that are spoken as weights & measures.  The space as a first character is important so I don't get every name that ends with one, only names that end with a separate letter.

If a match is found, convert it to a spoken version of the letter. I do understand that this can be a bit of a dirty conversion as it will depend on the voice used, but it will still be better than a totally wrong word (for me anyway).  If no match is found, then the raw name will be spoken, otherwise I put the original first half plus the new converted half together to be spoken.

I've tried this with the three default windows 10 voices (Hazel, David & Zira) and Ivona Amy (which is my primary voice).  Only Amy and Zira seem to use the letters as abbreviations.  Of course, this 'fix' is tuned to Amy, but all voices seem to be ok with my spellings, apart from the one for 'A'.  The three Win10 voices do have a hard time with that one.

I used this in a new script to test with:-
```
1 a .
ay.

1 g .
gee.

1 h .
aitch.

1 l .
elle.

1 m .
emm.

1 t .
tee.

1 y .
why.
```

**Comment by Darkcyde13** [comment]
I should also note that in my Body/Star Scanned events, I've set a state variable that I use to prevent the body/star names being spoken when the respective reports are called from them.  It was getting tedious hearing the long names of sectors being read out twice in quick succession.

This is why I've got both the reportbody.name and event.name in the same place as in my scripts, only one will ever be used at any one time.  If used from the default scripts, I believe they would both be spoken when a body is scanned.
